README.rst
============
Python Rules
============

:::

    context1 = {
        "price": {
            "value": 3.45,
            "unit": "USD"
        },
        "name": "foobar inc"
    }

    context2 = {
        "price": {
            "value": 45.00,
            "unit": "USD"
        },
        "name": "cyanic ltd"
    }


    from rules import expressions
    r = expressions.ContextReference("price.value")
    c = expressions.Constant(1.25)
    o = expressions.Operator("*", r, c)
    assert str(o) == "price.value * 1.25"

    assert o(context) == 4.3125

    assert o(context2) == 56.25

    expr = {
        "type": "operator",
        "operator": "*",
        "left_side": {
            "type": "context reference",
            "name": "price.value",
            "default": None,
        },
        "right_side": 1.25
    }

    expr_json = '''{
        "type": "operator",
        "operator": "*",
        "left_side": {
            "type": "context reference",
            "name": "price.value",
            "default": null
        },
        "right_side": 1.25
    }'''

    from rules import parser
    func1 = parser.parse(expr)
    func2 = parser.parse(expr_json)

    assert str(func1) == str(func2) == "price.value * 1.25"
    assert func1(context1) == func2(context1) == 4.3125

    assert func1(context2) == func2(context2) == 56.25

Features
--------

* TODO

Getting Started
---------------

Read the ``CONTRIBUTING.rst`` docs to get things set up.


CHANGELOG.md
# Changelog

## 0.1.0 (2020-04-14)

* Initial creation


tests/test_rules.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `rules` package."""
import pytest


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


rules/parser.py
import json
from rules import expressions


def parse_operator(expr):
    opr = expr["operator"]
    left_side = parse(expr["left_side"])
    right_side = parse(expr["right_side"])
    return expressions.Operator(opr, left_side, right_side)


def parse_context_ref(expr):
    name = expr["name"]
    default = expr.get("default", None)
    return expressions.ContextReference(name, default)


def parse_constant(expr):
    return expressions.Constant(expr)


def parse_expr(expr):
    parse_map = {
        "operator": parse_operator,
        "context reference": parse_context_ref,
        "constant": parse_constant,
    }

    if isinstance(expr, list):
        return None  # TODO

    if isinstance(expr, (str, bool, int, float)):
        return expressions.Constant(expr)

    if not isinstance(expr, dict):
        return None

    if "type" not in expr:
        return expressions.Constant(expr)

    expr_type = expr["type"]
    return parse_map[expr_type](expr)


def parse(expr):
    if isinstance(expr, str):
        return parse_expr(json.loads(expr))
    return parse_expr(expr)


rules/expressions.py
"""Main module."""
import operator
from abc import ABC, abstractmethod


class OperatorError(Exception):
    pass


class ExpressionObject(ABC):

    @abstractmethod
    def __str__(self) -> str:
        """
        A string representation of the expression

        Returns:
            str
        """
        return ""

    @abstractmethod
    def __call__(self, ctx):
        """
        Perform the operations based on the context and other arguments

        Args:
            ctx: The expression context

        Returns:
            Any
        """
        pass


class Constant(ExpressionObject):
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return str(self.value)

    def __call__(self, ctx):
        return self.value


class ContextReference(ExpressionObject):
    def __init__(self, name, default=None):
        self.name = name
        self.default = default

    def __str__(self):
        return self.name

    def resolve_name(self, ctx):
        """
        Get a key or attr ``name`` from obj or None

        Copied and modified from Django Template variable resolutions

        Args:
            ctx:

        Returns:
            Any or None
        """
        lookups = self.name.split(".")
        current = ctx
        try:  # catch-all for unexpected failures
            for bit in lookups:
                try:  # dictionary lookup
                    current = current[bit]
                    # ValueError/IndexError are for numpy.array lookup on
                    # numpy < 1.9 and 1.9+ respectively
                except (TypeError, AttributeError, KeyError, ValueError, IndexError):
                    try:  # attribute lookup
                        current = getattr(current, bit)
                    except (TypeError, AttributeError):
                        # Reraise if the exception was raised by a @property
                        if bit in dir(current):
                            raise
                        try:  # list-index lookup
                            current = current[int(bit)]
                        except (
                            IndexError,  # list index out of range
                            ValueError,  # invalid literal for int()
                            KeyError,  # current is a dict without `int(bit)` key
                            TypeError,
                        ):  # unsubscriptable object
                            return self.default
            return current
        except Exception:  # NOQA
            return self.default

    def __call__(self, ctx):
        return self.resolve_name(ctx)


class Operator(ExpressionObject):
    operator_map = {
        '+': operator.add,
        '-': operator.sub,
        '*': operator.mul,
        '/': operator.floordiv,
        '^': operator.pow,
        #
        '%': operator.mod,
        #
        '&': operator.and_,
        '|': operator.or_,
        #
        '==': operator.eq,
        '!=': operator.ne,
        '>': operator.gt,
        '>=': operator.ge,
        '<': operator.lt,
        '<=': operator.le,
        #
        'in': operator.contains,
    }

    def __init__(self, opr, left_side, right_side):
        self.left_side = left_side
        self.right_side = right_side
        self.opr = opr
        if opr not in self.operator_map.keys():
            op_list = ", ".join(self.operator_map.keys())
            raise OperatorError(f'"{opr}" is not a valid operator. Must be')

    def __str__(self):
        return f"{self.left_side} {self.opr} {self.right_side}"

    def __call__(self, ctx):
        return self.operator_map[self.opr](self.left_side(ctx), self.right_side(ctx))


rules/play.py
from rules import expressions, parser

context1 = {
    "price": {
        "value": 3.45,
        "unit": "USD"
    },
    "name": "foobar inc"
}

context2 = {
    "price": {
        "value": 45.00,
        "unit": "USD"
    },
    "name": "cyanic ltd"
}

expr = {
    "type": "operator",
    "operator": "*",
    "left_side": {
        "type": "context reference",
        "name": "price.value",
        "default": None,
    },
    "right_side": 1.25
}

expr_json = '''{
    "type": "operator",
    "operator": "*",
    "left_side": {
        "type": "context reference",
        "name": "price.value",
        "default": null
    },
    "right_side": 1.25
}'''

func1 = parser.parse(expr)
func2 = parser.parse(expr_json)

assert str(func1) == str(func2) == "price.value * 1.25"
assert func1(context1) == func2(context1) == 4.3125
assert func1(context2) == func2(context2) == 56.25

r = expressions.ContextReference("price.value")
c = expressions.Constant(1.25)
o = expressions.Operator("*", r, c)
assert str(o) == "price.value * 1.25"

assert o(context1) == 4.3125

assert o(context2) == 56.25
